netsize/netsize.go

```diff
@@ -0,0 +1,268 @@
+package netsize
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+	"github.com/libp2p/go-libp2p/core/peer"
+	ks "github.com/whyrusleeping/go-keyspace"
+)
+
+var (
+	ErrNotEnoughData   = fmt.Errorf("not enough data")
+	ErrWrongNumOfPeers = fmt.Errorf("expected bucket size number of peers")
+)
+
+var (
+	logger                   = logging.Logger("dht/netsize")
+	MaxMeasurementAge        = 2 * time.Hour
+	MinMeasurementsThreshold = 5
+	MaxMeasurementsThreshold = 150
+	keyspaceMaxInt, _        = new(big.Int).SetString(strings.Repeat("1", 256), 2)
+	keyspaceMaxFloat         = new(big.Float).SetInt(keyspaceMaxInt)
+)
+
+type Estimator struct {
+	localID    kbucket.ID
+	rt         *kbucket.RoutingTable
+	bucketSize int
+
+	measurementsLk sync.RWMutex
+	measurements   map[int][]measurement
+
+	netSizeCache *float64
```

Comment:
You don't really need a pointer here, you can use values like `0`, any negative number or `-Inf`, `Inf`. To signify the estimation is not valid.

Also do we really need float64 precision ? The estimations are multiple orders of magnitude more reliable than `+-1` precision, I would use an `uint64` and have `0` means not cached.
Then you can `atomic.LoadUint64` to have a lock free fast path when someone tries to fetch the network size.

The logic would look like this:
```go
func (e *Estimator) NetworkSize() (uint64, error) {
  if estimate := atomic.LoadUint64(&e.netSizeCache); estimation != 0 {
		logger.Debugw("Cached network size estimation", "estimate", estimate)
    return estimate, nil
  }

	e.measurementsLk.Lock()
	defer e.measurementsLk.Unlock()

  // Check a second time, this is needed because we maybe had to wait on an other goroutine doing the computation.
  // Then the computation was just finished by the other goroutine so need to redo it.
  if estimate := e.netSizeCache; estimate != 0 {
		logger.Debugw("Cached network size estimation", "estimate", estimate)
    return estimation, nil
  }
	
  // Compute netSize	...

  atomic.PutUint64(&e.netSizeCache, netSize)

	logger.Debugw("New network size estimation", "estimate", netSize)
	return netSize, nil
}
```

Reply:
decided to have `-1` as the invalid value and use `int32` instead of `uint32` because technically `0` is a valid network size. This only becomes a problem as soon as we have >2.1 billion peers